text/3721-homogeneous-try-blocks.md

```diff
@@ -0,0 +1,759 @@
+that occurs *between* the two `?`s.  We'd be happy for it to just be the same
+`io::Result<_>` as in the other places, but there's nothing saying it *must* be that.
+To the compiler, we might want some completely different error type that happens
+to support conversion to and from `io::Error`.
+
+The easiest fix here is the annotate the return type of the closure, as follows:
+
+```rust
+use std::io::{self, BufRead};
+pub fn concat_lines(reader: impl BufRead) -> io::Result<String> {
+    let mut out = String::new();
+    reader.lines().try_for_each(|line| -> io::Result<()> { // <-- return type
+        let line = line?;
+        out.push_str(&line);
+        Ok(())
+    })?;
+    Ok(out)
+}
+```
+
+<!-- https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a3870699c7df0df06eb720c036d94f0e -->
+
+But it would be nice to have a way to request that "the obvious thing" should happen.
+
+This same kind of problem happens with `try{}` blocks as they were implemented
+in nightly at the time of writing of this RFC.  The desugaring of `?` in a `try{}`
+block was essentially the same as in a function or closure, differing only in that
+it "returns" the value from the block instead of from the enclosing function.
+
+For example, this works great as it the type context available from the return type:
+
+```rust
+pub fn adding_a(x: Option<i32>, y: Option<i32>, z: Option<i32>) -> Option<i32> {
+    Some(x?.checked_add(y?)?.checked_add(z?)?)
+}
+```
+
+<!-- https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0e383fab261dc7a39adf61975a307af3 -->
+
+Suppose, however, that you wanted to do more in the method after the additions,
+and thus added a `try{}` block around it:
+
+```rust
+#![feature(try_blocks)]
+pub fn adding_b(x: Option<i32>, y: Option<i32>, z: Option<i32>) -> i32 {
+    try { // pre-RFC version
+        x?.checked_add(y?)?.checked_add(z?)?
+    }
+    .unwrap_or(0)
+}
+```
+
+<!-- https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e53dda532b44eecd6d1725143d04f900 -->
+
+That doesn't compile, since a (non-trait) method call required the type be determined:
+
+```text
+error[E0282]: type annotations needed
+ --> src/lib.rs:3:5
+  |
+3 | /     try { // pre-RFC version
+4 | |         x?.checked_add(y?)?.checked_add(z?)?
+5 | |     }
+  | |_____^ cannot infer type
+  |
+  = note: type must be known at this point
+```
+
+This is, in a way, more annoying than the `Result` case.  Since at least there,
+there's the possibility that one wants the `io::Error` converted into some
+`my_special::Error`.  But for `Option`, there's no conversion for `None`.
+While it's possible that there's some other type that accepts its residual,
+the normal case is definitely that it just stays a `None`.
+
+This RFC proposes using the unannotated `try { ... }` block as the marker to
+request a slightly-different `?` desugaring that stays in the same family.
```

Comment:
Given that the motivation section beings with closures, this feels like an unexpected segue. I was expecting it the RFC would now explain how the very first example, the closure, can be accepted. It would be good to explain somewhere why that is not an option, and we have to insert a `try` block (with changed desugaring) instead.